File: Services/API/Wordnik/Models/Response/PartOfSpeech.swift
//
//  PartOfSpeech.swift
//  My Dictionary
//
//  Created by Aleksandr Riakhin on 3/9/25.
//

import Foundation
import Core

enum PartOfSpeech: String, Decodable {
    case noun
    case adjective
    case verb
    case adverb
    case interjection
    case pronoun
    case preposition
    case abbreviation
    case affix
    case article
    case auxiliaryVerb = "auxiliary verb"
    case conjunction
    case definiteArticle = "definite article"
    case familyName = "family name"
    case givenName = "given name"
    case idiom
    case imperative
    case nounPlural = "noun plural"
    case nounPossessive = "noun posessive"
    case pastParticiple = "past participle"
    case phrasalPrefix = "phrasal prefix"
    case properNoun = "proper noun"
    case properNounPlural = "proper noun plural"
    case properNounPossessive = "proper noun posesessive"
    case suffix
    case verbIntransitive = "intransitive verb"
    case verbTransitive = "transitive verb"
    case unknown // Handles unexpected values

    /// Custom initializer that prevents decoding from failing.
    public init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        let rawValue = try container.decode(String.self)
        self = PartOfSpeech(rawValue: rawValue) ?? .unknown
    }

    var coreValue: Core.PartOfSpeech {
        switch self {
        case .adjective: .adjective
        case .adverb: .adverb
        case .pronoun: .pronoun
        case .preposition: .preposition
        case .verb, .auxiliaryVerb, .verbIntransitive, .verbTransitive: .verb
        case .conjunction: .conjunction
        case .noun, .nounPlural, .nounPossessive, .properNoun, .properNounPlural, .properNounPossessive, .familyName, .givenName: .noun
        case .interjection: .exclamation
        default: .unknown
        }
    }
}

File: MyDictionaryAndroid/app/src/main/java/com/dor/mydictionary/core/PartOfSpeech.kt
package com.dor.mydictionary.core

enum class PartOfSpeech(val rawValue: String) {
    Noun("noun"),
    Verb("verb"),
    Adjective("adjective"),
    Adverb("adverb"),
    Conjunction("conjunction"),
    Pronoun("pronoun"),
    Preposition("preposition"),
    Exclamation("exclamation"),
    Interjection("interjection"),
    Idiom("idiom"),
    Phrase("phrase"),
    Unknown("unknown");

    val isExpression: Boolean
        get() = this == Idiom || this == Phrase

    companion object {
        fun fromRawValue(value: String): PartOfSpeech {
            return entries.find { it.rawValue == value } ?: Unknown
        }
        
        val wordCases: List<PartOfSpeech>
            get() = entries.filter { !it.isExpression && it != Unknown }
        
        val expressionCases: List<PartOfSpeech>
            get() = listOf(Idiom, Phrase)
    }
}